File: organization/human_web/learning.js
"use strict";
function renderLearning(){
 const root=$('learning-list');root.replaceChildren();
 const choices=kind=>values(kind).map(x=>[x.id,x.id+' · '+(x.status||x.outcome||'')]);
 const none=kind=>[['none','无 / 新建版本系列'],...choices(kind)];
 function form(title,cmd,fields,transform=x=>x){const d=node('details',undefined,root);node('summary',title,d);const inputs={};for(const [key,label,kind,options] of fields){inputs[key]=kind==='select'?select(d,label,options):field(d,label,'',kind==='text'?'textarea':'input');}button(d,'记录：'+title,()=>{const data={};for(const [key,label,kind]of fields){let value=inputs[key].value.trim();if(!value)throw Error('请填写：'+label);data[key]=kind==='list'?value.split(',').map(x=>x.trim()).filter(Boolean):value;}if(!['assemble_route'].includes(cmd))data.id=crypto.randomUUID();return command(cmd,transform(data));});}
 form('1 · 提出候选认知','propose_claim',[
 ['return_id','来源 Return','select',choices('Return')],['reviewer','独立验证 Human','select',choices('Human')],['previous_id','修订自哪个知识版本','select',none('KnowledgeRevision')],['core_claim','核心主张','text'],['scope','适用情境','text'],['boundary','所需操作边界（逗号分隔）','list'],['mechanism','候选机制解释','text'],['transfer_conditions','迁移条件与不适用情形','text'],['task_ids','本轮适用 Task 编号（逗号分隔）','list']]);
 form('2 · 关联证据','capture_evidence',[
 ['claim_id','对应候选主张','select',choices('LearningClaim')],['return_id','证据来源 Return','select',choices('Return')],['relation','与主张的关系','select',[['inconclusive','暂无法判断'],['supports','支持'],['contradicts','反驳'],['discriminates','区分其他解释']]],['rationale','证据关系及理由','text'],['upstream_sources','共同上游来源标识（逗号分隔，必须如实填写）','list']]);
 form('3 · 反证与独立验证','verify_claim',[
 ['claim_id','待验证主张','select',choices('LearningClaim')],['decision','判断','select',[['rejected','证据不足 / 拒绝'],['validated','在声明范围内确认']]],['evidence_ids','所有相关 Evidence 编号（逗号分隔）','list'],['falsification','反证尝试及观察结果','text'],['alternatives','替代解释','text'],['limitations','局限与未验证范围','text'],['independence','证据独立性判断依据','text']]);
 form('4 · 整理候选能力包','assemble_route',[
 ['package_id','已捕获候选路线','select',choices('CapabilityRoute')],['knowledge_id','依赖的知识版本','select',choices('KnowledgeRevision')],...['data','tools','agent_configuration','runtime','practice','oversight'].map((k,i)=>[k,['数据及版本','工具及版本','Agent 配置','运行环境','可复现方法','Human 审查与监督需求'][i],'text'])],d=>{const components={};for(const k of ['data','tools','agent_configuration','runtime','practice','oversight']){components[k]=d[k];delete d[k];}return {...d,components};});
 form('5 · 记录执行中的采用或拒绝','use_knowledge',[
 ['execution_id','当前 Execution','select',choices('Execution')],['knowledge_id','知识版本','select',choices('KnowledgeRevision')],['package_id','能力包（可不使用）','select',none('CapabilityRoute')],['decision','使用决定','select',[['rejected','拒绝采用'],['adopted','采用']]],['reason','范围判断与使用理由','text']]);
 form('6 · 关联后续结果','record_learning_outcome',[
 ['use_id','使用记录','select',choices('KnowledgeUse')],['return_id','同一 Execution 的 Return','select',choices('Return')],['interpretation','结果观察与解释（不默认归因于知识）','text']]);
 form('7 · 核对另一 HAU 的复现','record_reproduction',[
 ['package_id','能力包','select',choices('CapabilityRoute')],['use_id','采用记录','select',choices('KnowledgeUse')],['return_id','复现 Return','select',choices('Return')]]);
 for(const [kind,title]of [['LearningClaim','候选解释'],['LearningEvidence','证据及来源快照'],['LearningVerification','独立验证记录'],['KnowledgeRevision','知识版本与变化'],['CapabilityRoute','候选路线与复现状态'],['KnowledgeUse','采用 / 拒绝'],['LearningOutcome','结果反馈'],['CapabilityReproduction','跨 HAU 复现']]){const c=card(root,title+' · '+values(kind).length);for(const x of values(kind))dump(c,x.id+' · '+(x.status||x.decision||x.relation||x.outcome||''),x);}
}

File: organization/learning.py
"""Bounded, explicit learning records. No automatic truth or policy promotion."""
from organization.core import require
from organization.store import DomainError


def items(s, kind):
    return list(s.state.get(kind, {}).values())


def text(value):
    if not isinstance(value, str) or not value.strip():
        raise DomainError('Nonempty explanation required')


def source(s, rid):
    r=s.get('Return',rid);e=s.get('Execution',r['execution_id'])
    return r,e


def accepted(s,rid):
    return any(a['return_id']==rid and a['accepted'] for a in items(s,'Acceptance'))


def invalidate(s,kid,auth,status):
    k=s.get('KnowledgeRevision',kid);k['status']=status
    s.emit('KnowledgeRevision',k,'KnowledgeSuperseded' if status=='superseded' else 'KnowledgeChallenged',auth,k['author'])
    for route in items(s,'CapabilityRoute'):
        if route['knowledge_id']==kid and route['status']!='stale':
            route=dict(route,status='stale')
            s.emit('CapabilityRoute',route,'RouteStale',auth,route['owner'])


def capture_route(s,r,e):
    """Capture candidate within the formal success event; no truth/authority grant."""
    route={'id':'route/'+r['id'],'source_return':r['id'],'source_hau':e['hau_id'],
           'owner':e['accountable_owner'],'knowledge_id':'none','status':'candidate','reproduction_ids':[],
           'components':{'observed_terrain':r['observed_terrain'],'practice':r['major_execution_facts'],
                         'reflection':r['reflection'],'missing':['knowledge','data','tools','agent_configuration','runtime','oversight']}}
    s.attach('CapabilityRoute',route)


def handle(s,command,d):
    if command=='propose_claim':
        require(d,['id','return_id','reviewer','previous_id','core_claim','scope','boundary','mechanism','transfer_conditions','task_ids'])
        s.new('LearningClaim',d['id']);r,e=source(s,d['return_id'])
        auth=s.authority(e['accountable_owner']);s.human(d['reviewer'])
        if d['reviewer']==s.actor:raise DomainError('Separate Human verifier required',403)
        for f in ('core_claim','scope','mechanism','transfer_conditions'):text(d[f])
        if not d['task_ids'] or not isinstance(d['task_ids'],list):raise DomainError('Explicit candidate application tasks required')
        for tid in d['task_ids']:s.get('Task',tid)
        if d['previous_id']!='none':
            old=s.get('KnowledgeRevision',d['previous_id']);s.authority(old['author'])
            if old['status']=='superseded':raise DomainError('Revise current or challenged revision only',409)
        obj={k:v for k,v in d.items() if k!='return_id'}
        obj.update(author=s.actor,source_return=d['return_id'],status='candidate')
        s.emit('LearningClaim',obj,'ClaimProposed',auth,s.actor)
    elif command=='capture_evidence':
        require(d,['id','claim_id','return_id','relation','rationale','upstream_sources'])
        s.new('LearningEvidence',d['id']);claim=s.get('LearningClaim',d['claim_id']);r,e=source(s,d['return_id'])
        auth=s.authority(e['accountable_owner']);text(d['rationale'])
        if d['relation'] not in ('supports','contradicts','discriminates','inconclusive'):raise DomainError('Unknown evidence relation')
        if not isinstance(d['upstream_sources'],list) or not d['upstream_sources'] or any(not isinstance(x,str) or not x.strip() for x in d['upstream_sources']):raise DomainError('Explicit upstream provenance required')
        if any(x['claim_id']==d['claim_id'] and x['return_id']==d['return_id'] for x in items(s,'LearningEvidence')):raise DomainError('Same Return is one source for this claim',409)
        snapshot={'return':r,'execution':e,'task':s.get('Task',e['task_id']),'reviews':[x for x in items(s,'Review') if x['return_id']==r['id']],'acceptances':[x for x in items(s,'Acceptance') if x['return_id']==r['id']]}
        s.emit('LearningEvidence',{**d,'owner':s.actor,'snapshot':snapshot},'EvidenceLinked',auth,s.actor,e['task_id'])
        routeid='route/'+r['id']
        if r['outcome']=='success' and routeid not in s.state.get('CapabilityRoute',{}):
            route={'id':routeid,'source_return':r['id'],'source_hau':e['hau_id'],'owner':s.actor,'knowledge_id':'none','status':'candidate','reproduction_ids':[],
                   'components':{'observed_terrain':r['observed_terrain'],'practice':r['major_execution_facts'],'reflection':r['reflection'],'review_refs':[x['id'] for x in snapshot['reviews']], 'missing':['knowledge','data','tools','agent_configuration','runtime','oversight']}}
            s.emit('CapabilityRoute',route,'RouteCaptured',auth,s.actor,e['task_id'])
        if d['relation']=='contradicts':
            for k in items(s,'KnowledgeRevision'):
                if k['claim_id']==claim['id'] and k['status']=='validated':invalidate(s,k['id'],auth,'challenged')
    elif command=='verify_claim':
        require(d,['id','claim_id','decision','evidence_ids','falsification','alternatives','limitations','independence'])
        s.new('LearningVerification',d['id']);c=s.get('LearningClaim',d['claim_id']);auth=s.authority(c['reviewer'])
        if c['status']!='candidate':raise DomainError('Claim already decided; propose a revision',409)
        if d['decision'] not in ('validated','rejected'):raise DomainError('Unknown verification decision')
        for f in ('falsification','alternatives','limitations','independence'):text(d[f])
        evidence=[s.get('LearningEvidence',eid) for eid in d['evidence_ids']]
        if not evidence or any(x['claim_id']!=c['id'] for x in evidence):raise DomainError('Evidence must belong to this claim')
        if set(d['evidence_ids'])!={x['id'] for x in items(s,'LearningEvidence') if x['claim_id']==c['id']}:raise DomainError('Verification must consider all current evidence',409)
        if d['decision']=='validated':
            if any(x['relation']=='contradicts' for x in evidence):raise DomainError('Counter evidence requires a revised bounded claim',409)
            support=[x for x in evidence if x['relation']=='supports']
            if len({x['snapshot']['execution']['hau_id'] for x in support})<2:raise DomainError('Minimum local gate: two source HAUs required',409)
            upstream=set()
            for x in support:
                if upstream.intersection(x['upstream_sources']):raise DomainError('Shared upstream cannot count as independent evidence',409)
                upstream.update(x['upstream_sources'])
                if not accepted(s,x['return_id']):raise DomainError('Supporting Return requires separate review/acceptance',409)
        c['status']=d['decision'];s.emit('LearningVerification',{**d,'reviewer':s.actor},'ClaimVerified',auth,c['author']);s.attach('LearningClaim',c)
        if d['decision']=='validated':
            previous=c['previous_id'];old=s.get('KnowledgeRevision',previous) if previous!='none' else None
            if old and old['status']=='superseded':raise DomainError('Revision target changed; rebase candidate explicitly',409)
            fields=('core_claim','scope','boundary','mechanism','transfer_conditions','task_ids')
            delta={f:{'before':old[f] if old else None,'after':c[f]} for f in fields if not old or old[f]!=c[f]}
            k={f:c[f] for f in fields};k.update(id=c['id'],claim_id=c['id'],author=c['author'],previous_id=previous,status='validated',version=old['version']+1 if old else 1,evidence_ids=d['evidence_ids'],delta=delta)
            if old:invalidate(s,old['id'],auth,'superseded')
            s.emit('KnowledgeRevision',k,'KnowledgePublished',auth,c['author'])
    elif command=='assemble_route':
        require(d,['package_id','knowledge_id','components'])
        route=s.get('CapabilityRoute',d['package_id']);auth=s.authority(route['owner']);k=s.get('KnowledgeRevision',d['knowledge_id'])
        if route['knowledge_id']!='none':raise DomainError('Package binding immutable; capture a new route for changed dependencies',409)
        if k['status']!='validated':raise DomainError('Validated bounded knowledge required',409)
        if not any(s.get('LearningEvidence',eid)['return_id']==route['source_return'] for eid in k['evidence_ids']):raise DomainError('Route must be grounded in this knowledge evidence',409)
        require(d['components'],['data','tools','agent_configuration','runtime','practice','oversight'])
        for value in d['components'].values():text(value)
        route.update(knowledge_id=k['id'],components=d['components'])
        s.emit('CapabilityRoute',route,'RouteAssembled',auth,route['owner'])
    elif command=='use_knowledge':
        require(d,['id','execution_id','knowledge_id','decision','reason','package_id'])
        s.new('KnowledgeUse',d['id']);e=s.get('Execution',d['execution_id']);auth=s.executor(e,command);k=s.get('KnowledgeRevision',d['knowledge_id'])
        if e['status'] not in ('claimed','running','blocked'):raise DomainError('Knowledge use requires live Execution',409)
        if k['status']!='validated' or e['task_id'] not in k['task_ids']:raise DomainError('Knowledge not valid for explicit task scope',403)
        if not set(k['boundary']).issubset(e['runtime_boundary']):raise DomainError('Knowledge cannot expand runtime boundary',403)
        if d['decision'] not in ('adopted','rejected'):raise DomainError('Explicit adoption or rejection required')
        text(d['reason'])
        if d['package_id']!='none':
            route=s.get('CapabilityRoute',d['package_id'])
            if route['status']=='stale' or route['knowledge_id']!=k['id']:raise DomainError('Package dependency not valid',409)
        s.emit('KnowledgeUse',d,'KnowledgeUsed',auth,e['accountable_owner'],e['task_id'])
    elif command=='record_learning_outcome':
        require(d,['id','use_id','return_id','interpretation'])
        s.new('LearningOutcome',d['id']);u=s.get('KnowledgeUse',d['use_id']);r,e=source(s,d['return_id']);auth=s.authority(e['accountable_owner'])
        if e['id']!=u['execution_id']:raise DomainError('Outcome must belong to the usage Execution',409)
        if any(x['use_id']==u['id'] for x in items(s,'LearningOutcome')):raise DomainError('Outcome already linked',409)
        text(d['interpretation']);s.emit('LearningOutcome',{**d,'owner':s.actor},'LearningOutcomeRecorded',auth,s.actor,e['task_id'])
        # Failure is a revalidation signal, not proof that knowledge caused it.
        if r['outcome']!='success':
            k=s.get('KnowledgeRevision',u['knowledge_id'])
            if k['status']=='validated':invalidate(s,k['id'],auth,'challenged')
    elif command=='record_reproduction':
        require(d,['id','package_id','use_id','return_id'])
        s.new('CapabilityReproduction',d['id']);route=s.get('CapabilityRoute',d['package_id']);auth=s.authority(route['owner']);u=s.get('KnowledgeUse',d['use_id']);r,e=source(s,d['return_id']);k=s.get('KnowledgeRevision',route['knowledge_id'])
        if route['status']=='stale' or k['status']!='validated':raise DomainError('Dependency requires revalidation',409)
        if u['package_id']!=route['id'] or u['execution_id']!=e['id'] or u['decision']!='adopted':raise DomainError('Reproduction requires explicit package adoption',409)
        if e['hau_id']==route['source_hau']:raise DomainError('Another HAU must reproduce',409)
        if any(x['use_id']==u['id'] for x in items(s,'CapabilityReproduction')):raise DomainError('Reproduction already recorded',409)
        if not any(x['use_id']==u['id'] and x['return_id']==r['id'] for x in items(s,'LearningOutcome')):raise DomainError('Link observed outcome first',409)
        if not any(a['return_id']==r['id'] for a in items(s,'Acceptance')):raise DomainError('Reproduction needs a separate final Acceptance decision',409)
        outcome='reproduced' if r['outcome']=='success' and accepted(s,r['id']) else 'not_reproduced'
        s.emit('CapabilityReproduction',{**d,'reviewer':s.actor,'outcome':outcome},'ReproductionRecorded',auth,route['owner'])
        route['reproduction_ids'].append(d['id']);route['status']=outcome if outcome=='reproduced' else 'candidate';s.attach('CapabilityRoute',route)
